### PR TITLE
fix(database): allow both metadata and annotations

### LINKED ISSUE #14405: [Bug]: unexpected results when using metadata and annotations in elgg_get_entities()
### Elgg version

5.0

### Source of your Elgg installation

GIT clone

### Describe the bug

When using both metadata and annotation options in elgg_get_entities() the results don't match the expectation
```php
// get faked entities with likes
$result = elgg_get_entities([
   'metadata_names' => '__faker',
   'annotation_name_value_pairs' => [
       [
          'name' => 'likes',
          'value' => 'likes',
          'case_sensitive' => false,
          'type' => ELGG_VALUE_STRING,
      ],
   ],
]);
```

this will result in no entities as both the metadata and annotations joins are based on the name `n_table` which is already joined so no annotations join will be added

https://github.com/Elgg/Elgg/blob/bd206c1491fe9b34fa979f55b97b063070ea3ec8/engine/classes/Elgg/Database/Entities.php#L268-L284

https://github.com/Elgg/Elgg/blob/bd206c1491fe9b34fa979f55b97b063070ea3ec8/engine/classes/Elgg/Database/Entities.php#L296-L310

### Steps to reproduce

See code example in description

### Expected behaviour

Create the correct number of joins, since this will probably involve renaming the joins this could cause a lot of issues (order by's, etc)

### Additional notes

Search in the codebase for `n_table`

### PHP version

_No response_

### PHP error log

_No response_

### PR DESCRIPTION
**Pull request by jeabakker**
fixes #14405

### CODE REVIEW

File: engine/classes/Elgg/Database/Repository.php [hunk 1]
```diff
@@ -181,18 +181,18 @@ public function select($expression) {
 	 * @param string $joined_table   Name of the table (with or without dbprefix)
 	 * @param string $joined_alias   Alias of the joined table
 	 *                               If not set, the alias will be assigned automatically
-	 * @param string $x              Base column, e.g. 'n_table.entity_guid'
+	 * @param string $join_column    Base column, e.g. 'n_table.entity_guid'
 	 *                               This value is NOT a query parameter and will not be sanitized
 	 * @param string $comparison     Comparison operator, e.g. '=', 'not like' etc
 	 * @param mixed  $y              Comparison value(s)
```
Comment (jdalsem) [inline engine/classes/Elgg/Database/Repository.php:187]: if changing x why not y

File: engine/classes/Elgg/Database/River.php [hunk 2]
```diff
@@ -178,7 +178,7 @@ public function get($limit = null, $offset = null, $callback = null) {
 	/**
 	 * Execute the query resolving calculation, count and/or batch options
 	 *
-	 * @return array|\ElggData[]|\ElggEntity[]|false|int
+	 * @return array|\ElggData[]|\ElggEntity[]|int|\ElggBatch
```
Comment (jdalsem) [inline engine/classes/Elgg/Database/River.php:181]: check

File: engine/classes/Elgg/Upgrades/NotificationsPrefix.php [hunk 3]
```diff
@@ -59,27 +60,27 @@ public function run(Result $result, $offset): Result {
...
+			$select = Select::fromTable(RelationshipsTable::TABLE_NAME, 'r1');
 			
 			// exclude already migrated relationships
-			$exists = $select->subquery('entity_relationships', 'r2');
+			$exists = $select->subquery(RelationshipsTable::TABLE_NAME, 'r2');
 			$exists->select('1')
-				->where($select->compare('r1.guid_one', '=', 'r2.guid_one'))
-				->andWhere($select->compare('r1.guid_two', '=', 'r2.guid_two'))
-				->andWhere($select->compare('r2.relationship', '=', "{$relationship_prefix}:{$method}", ELGG_VALUE_STRING));
+				->where($select->compare("{$select->getTableAlias()}.guid_one", '=', 'r2.guid_one'))
```
Comment (jdalsem) [inline engine/classes/Elgg/Upgrades/NotificationsPrefix.php:68]: r2 alias (2 lines)

File: engine/classes/ElggSite.php [hunk 4]
```diff
@@ -57,9 +59,9 @@ public function getType(): string {
 	 */
 	public function save(): bool {
 		$db = $this->getDatabase();
```
Comment (jdalsem) [inline engine/classes/ElggSite.php:61]: no need to put $db in a variable

File: engine/tests/phpunit/integration/Elgg/Integration/ElggCoreGetEntitiesFromAnnotationsTest.php [hunk 5]
```diff
@@ -372,7 +374,7 @@ public function testElggGetEntitiesFromAnnotationsCalculateConstrainedByWhere()
 			'annotation_values' => array_unique(call_user_func_array('array_merge', $values)),
 			'annotation_sort_by_calculation' => 'sum',
 			'wheres' => [
-				new WhereClause("CAST(n_table.value as DECIMAL(10, 2)) > 0"),
+				new WhereClause("CAST(a_table.value as DECIMAL(10, 2)) > 0"),
```
Comment (jdalsem) [inline engine/tests/phpunit/integration/Elgg/Integration/ElggCoreGetEntitiesFromAnnotationsTest.php:377]: why not const?

File: engine/tests/phpunit/integration/Elgg/Integration/ElggCoreGetEntitiesFromAnnotationsTest.php [hunk 6]
```diff
@@ -391,7 +393,7 @@ public function testElggGetEntitiesFromAnnotationsCalculateConstrainedByWhere()
 			$annotations = $e->getAnnotations([
 				'annotation_name' => $name,
 				'where' => [
-					new WhereClause("CAST(n_table.value AS DECIMAL(10, 2)) > 0"),
+					new WhereClause("CAST(a_table.value AS DECIMAL(10, 2)) > 0"),
```
Comment (jdalsem) [inline engine/tests/phpunit/integration/Elgg/Integration/ElggCoreGetEntitiesFromAnnotationsTest.php:396]: const?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 7]
```diff
@@ -15,12 +15,12 @@
...
-		$select = Select::fromTable('annotations', 'n_table');
-		$select->select('COUNT(DISTINCT n_table.id) AS total');
+		$select = Select::fromTable(AnnotationsTable::TABLE_NAME, AnnotationsTable::DEFAULT_JOIN_ALIAS);
+		$select->select("COUNT(DISTINCT {$select->getTableAlias()}.id) AS total");
 
-		$select->addClause(new AnnotationWhereClause(), 'n_table');
+		$select->addClause(new AnnotationWhereClause(), $select->getTableAlias());
 
-		$select->join('n_table', 'entities', 'e', 'e.guid = n_table.entity_guid');
+		$select->join($select->getTableAlias(), EntityTable::TABLE_NAME, 'e', "e.guid = {$select->getTableAlias()}.entity_guid");
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:23]: should `e` alias come from a const?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 8]
```diff
@@ -55,17 +55,17 @@ public function testThrowsOnBadDataFormat() {
...
-		$select = Select::fromTable('annotations', 'n_table');
-		$select->select('DISTINCT n_table.*');
+		$select = Select::fromTable(AnnotationsTable::TABLE_NAME, AnnotationsTable::DEFAULT_JOIN_ALIAS);
+		$select->select("DISTINCT {$select->getTableAlias()}.*");
 
-		$select->addClause(new AnnotationWhereClause(), 'n_table');
+		$select->addClause(new AnnotationWhereClause(), $select->getTableAlias());
 
-		$select->join('n_table', 'entities', 'e', 'e.guid = n_table.entity_guid');
+		$select->join($select->getTableAlias(), EntityTable::TABLE_NAME, 'e', "e.guid = {$select->getTableAlias()}.entity_guid");
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:63]: same

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 9]
```diff
@@ -80,7 +80,7 @@ public function testCanExecuteGet() {
 			'offset' => 5,
 			'callback' => false,
 			'order_by' => [
-				new OrderByClause('n_table.id', 'ASC'),
+				new OrderByClause('a_table.id', 'ASC'),
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:83]: const?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 10]
```diff
@@ -125,27 +125,27 @@ public function testCanExecuteGetWithClauses() {
...
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:128]: const?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 11]
```diff
@@ -125,27 +125,27 @@ public function testCanExecuteGetWithClauses() {
...
 			],
 			'selects' => [
 				'max(e.time_created) AS newest',
 			],
 			'group_by' => [
-				'n_table.entity_guid',
+				'a_table.entity_guid',
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:134]: const?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 12]
```diff
@@ -160,17 +160,17 @@ function (QueryBuilder $qb) {
...
-		$select = Select::fromTable('annotations', 'n_table');
-		$select->select('DISTINCT n_table.*');
+		$select = Select::fromTable(AnnotationsTable::TABLE_NAME, AnnotationsTable::DEFAULT_JOIN_ALIAS);
+		$select->select("DISTINCT {$select->getTableAlias()}.*");
 
-		$select->addClause(new AnnotationWhereClause(), 'n_table');
+		$select->addClause(new AnnotationWhereClause(), $select->getTableAlias());
 
-		$select->join('n_table', 'entities', 'e', 'e.guid = n_table.entity_guid');
+		$select->join($select->getTableAlias(), EntityTable::TABLE_NAME, 'e', "e.guid = {$select->getTableAlias()}.entity_guid");
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:168]: const e?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 13]
```diff
@@ -185,7 +185,7 @@ public function testCanExecuteBatchGet() {
 			'offset' => 5,
 			'callback' => false,
 			'order_by' => [
-				new OrderByClause('n_table.id', 'ASC'),
+				new OrderByClause('a_table.id', 'ASC'),
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:188]: const?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 14]
```diff
@@ -208,17 +208,16 @@ public function testCanExecuteBatchGet() {
...
+		$select = Select::fromTable(AnnotationsTable::TABLE_NAME, AnnotationsTable::DEFAULT_JOIN_ALIAS);
+		$select->select("min({$select->getTableAlias()}.value) AS calculation");
 
 		$annotation = new AnnotationWhereClause();
 		$annotation->names = $annotation_names;
-		$select->addClause($annotation, 'n_table');
+		$select->addClause($annotation, $select->getTableAlias());
 
-		$select->join('n_table', 'entities', 'e', 'e.guid = n_table.entity_guid');
+		$select->join($select->getTableAlias(), EntityTable::TABLE_NAME, 'e', "e.guid = {$select->getTableAlias()}.entity_guid");
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:220]: const e?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 15]
```diff
@@ -208,17 +208,16 @@ public function testCanExecuteBatchGet() {
...
 		$select->addClause(new EntityWhereClause(), 'e');
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:221]: const?

File: engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php [hunk 16]
```diff
@@ -246,20 +245,19 @@ public function testCanExecuteAnnotationCalculation() {
...
+		$alias = $select->joinAnnotationTable($select->getTableAlias(), 'entity_guid', $annotation_calculation_names);
+		$select->select("min({$alias}.value) AS calculation");
 
 		$annotation = new AnnotationWhereClause();
 		$annotation->names = $annotation_names;
-		$select->addClause($annotation, 'n_table');
+		$select->addClause($annotation, $select->getTableAlias());
 
-		$select->join('n_table', 'entities', 'e', 'e.guid = n_table.entity_guid');
+		$select->join($select->getTableAlias(), EntityTable::TABLE_NAME, 'e', "e.guid = {$select->getTableAlias()}.entity_guid");
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/AnnotationsUnitTest.php:260]: const? maybe search for `'e'` in whole file...

File: engine/tests/phpunit/unit/Elgg/Database/Clauses/AccessWhereClauseUnitTest.php [hunk 17]
```diff
@@ -2,15 +2,15 @@
 
 namespace Elgg\Database\Clauses;
 
+use Elgg\Database\EntityTable;
 use Elgg\Database\QueryBuilder;
 use Elgg\Database\Select;
 use Elgg\UnitTestCase;
-use ElggUser;
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/Clauses/AccessWhereClauseUnitTest.php:8]: maybe (other PR) do a pass for all unwanted root classes with a use...

File: engine/tests/phpunit/unit/Elgg/Database/Clauses/WhereClauseUnitTest.php [hunk 18]
```diff
@@ -15,33 +16,33 @@ class WhereClauseUnitTest extends UnitTestCase {
...
+		$actual = $query->prepare($qb, $qb->getTableAlias());
 
 		$this->assertEquals($expected, $actual);
 		$this->assertEquals($this->qb->getParameters(), $qb->getParameters());
 	}
 
 	public function testBuildWhereClauseFromClosure() {
 
-		$expected = $this->qb->compare('alias.guid', '=', 25, ELGG_VALUE_INTEGER);
+		$expected = $this->qb->compare("{{$this->qb->getTableAlias()}}.guid", '=', 25, ELGG_VALUE_INTEGER);
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/Clauses/WhereClauseUnitTest.php:37]: too much accolades

File: engine/tests/phpunit/unit/Elgg/Database/Clauses/WhereClauseUnitTest.php [hunk 19]
```diff
@@ -15,33 +16,33 @@ class WhereClauseUnitTest extends UnitTestCase {
...
 
 		$closure = function (QueryBuilder $qb) {
-			return $qb->compare('alias.guid', '=', 25, ELGG_VALUE_INTEGER);
+			return $qb->compare("{{$this->qb->getTableAlias()}}.guid", '=', 25, ELGG_VALUE_INTEGER);
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/Clauses/WhereClauseUnitTest.php:40]: too much accolades

File: engine/tests/phpunit/unit/Elgg/Database/MetadataUnitTest.php [hunk 20]
```diff
@@ -15,10 +15,10 @@
 class MetadataUnitTest extends UnitTestCase {
 
 	public function testCanExecuteCount() {
-		$select = Select::fromTable('metadata', 'n_table');
-		$select->select('COUNT(DISTINCT n_table.id) AS total');
+		$select = Select::fromTable(MetadataTable::TABLE_NAME, MetadataTable::DEFAULT_JOIN_ALIAS);
+		$select->select("COUNT(DISTINCT {$select->getTableAlias()}.id) AS total");
 
-		$select->join('n_table', 'entities', 'e', 'e.guid = n_table.entity_guid');
+		$select->join($select->getTableAlias(), 'entities', 'e', "e.guid = {$select->getTableAlias()}.entity_guid");
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/MetadataUnitTest.php:21]: const entiteis and const e?

File: engine/tests/phpunit/unit/Elgg/Database/MetadataUnitTest.php [hunk 21]
```diff
@@ -90,22 +90,22 @@ public function testCanExecuteGet() {
...
-		$select->where($select->compare("$alias.value", 'IN', ['draft'], ELGG_VALUE_STRING));
+		$select->groupBy("{$select->getTableAlias()}.entity_guid");
+		$select->join($select->getTableAlias(), AnnotationsTable::TABLE_NAME, 'an', "{$select->getTableAlias()}.entity_guid = an.entity_guid");
+		$alias = $select->joinMetadataTable($select->getTableAlias(), 'entity_guid', 'status');
+		$select->where($select->compare("{$alias}.value", 'IN', ['draft'], ELGG_VALUE_STRING));
 		$select->having($select->compare('e.time_updated', 'IS NOT NULL'));
 
 
-		$select->join('n_table', 'entities', 'e', 'e.guid = n_table.entity_guid');
+		$select->join($select->getTableAlias(), EntityTable::TABLE_NAME, 'e', "e.guid = {$select->getTableAlias()}.entity_guid");
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/MetadataUnitTest.php:103]: ha... a partial const :)

File: engine/tests/phpunit/unit/Elgg/Database/RelationshipsUnitTest.php [hunk 22]
```diff
@@ -55,17 +55,17 @@ public function testThrowsOnBadDataFormat() {
...
-		$select = Select::fromTable('entity_relationships', 'er');
-		$select->select('DISTINCT er.*');
+		$select = Select::fromTable(RelationshipsTable::TABLE_NAME, 'er');
+		$select->select("DISTINCT {$select->getTableAlias()}.*");
 
-		$select->addClause(new RelationshipWhereClause(), 'er');
+		$select->addClause(new RelationshipWhereClause(), $select->getTableAlias());
 
-		$select->joinEntitiesTable('er', 'guid_one', 'inner', 'e');
+		$select->joinEntitiesTable($select->getTableAlias(), 'guid_one', 'inner', 'e');
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/RelationshipsUnitTest.php:63]: const e?

File: engine/tests/phpunit/unit/Elgg/Database/RelationshipsUnitTest.php [hunk 23]
```diff
@@ -94,20 +94,20 @@ public function testCanExecuteGet() {
...
 	public function testCanExecuteGetWithClauses() {
-		$select = Select::fromTable('entity_relationships', 'er');
-		$select->select('DISTINCT er.*');
+		$select = Select::fromTable(RelationshipsTable::TABLE_NAME, 'er');
+		$select->select("DISTINCT {$select->getTableAlias()}.*");
 		$select->addSelect('max(e.time_created) AS newest');
-		$select->groupBy('er.guid_one');
-		$select->join('er', 'metadata', 'n_table', 'er.guid_one = n_table.entity_guid');
+		$select->groupBy("{$select->getTableAlias()}.guid_one");
+		$select->join($select->getTableAlias(), MetadataTable::TABLE_NAME, 'n_table', "{$select->getTableAlias()}.guid_one = n_table.entity_guid");
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/RelationshipsUnitTest.php:101]: const n_table?

File: engine/tests/phpunit/unit/Elgg/Database/RelationshipsUnitTest.php [hunk 24]
```diff
@@ -202,20 +202,19 @@ public function testCanExecuteBatchGet() {
 	}
 
 	public function testCanExecuteAnnotationCalculation() {
-
 		$annotation_names = ['foo'];
 
-		$select = Select::fromTable('entity_relationships', 'er');
-		$select->select("min(n_table.value) AS calculation");
+		$select = Select::fromTable(RelationshipsTable::TABLE_NAME, 'er');
+		$select->select('min(a_table.value) AS calculation');
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/Database/RelationshipsUnitTest.php:208]: no const a_table?

File: mod/blog/views/default/blog/sidebar/revisions.php [hunk 25]
```diff
@@ -19,8 +19,8 @@
 $revisions = $blog->getAnnotations([
 	'annotation_name' => 'blog_revision',
 	'order_by' => [
-		new OrderByClause('n_table.time_created', 'DESC'),
-		new OrderByClause('n_table.id', 'DESC'),
+		new OrderByClause('a_table.time_created', 'DESC'),
```
Comment (jdalsem) [inline mod/blog/views/default/blog/sidebar/revisions.php:22]: was there no sort_by alternative approach?

File: mod/likes/views/default/likes/popup.php [hunk 26]
```diff
@@ -14,7 +14,7 @@
 	'annotation_name' => 'likes',
 	'limit' => 99,
 	'pagination' => false,
-	'order_by' => new OrderByClause('n_table.time_created', 'desc'),
+	'order_by' => new OrderByClause('a_table.time_created', 'desc'),
```
Comment (jdalsem) [inline mod/likes/views/default/likes/popup.php:17]: sort by?

File: mod/messageboard/actions/messageboard/add.php [hunk 27]
```diff
@@ -24,8 +24,8 @@
 	'guid' => $owner->guid,
 	'pagination' => false,
 	'order_by' => [
-		new OrderByClause('n_table.time_created', 'DESC'),
-		new OrderByClause('n_table.id', 'DESC'),
+		new OrderByClause('a_table.time_created', 'DESC'),
```
Comment (jdalsem) [inline mod/messageboard/actions/messageboard/add.php:27]: sort by?

File: mod/messageboard/views/default/resources/messageboard/owner.php [hunk 28]
```diff
@@ -18,8 +18,8 @@
 	'annotation_name' => 'messageboard',
 	'guid' => $page_owner->guid,
 	'order_by' => [
-		new OrderByClause('n_table.time_created', 'DESC'),
-		new OrderByClause('n_table.id', 'DESC'),
+		new OrderByClause('a_table.time_created', 'DESC'),
```
Comment (jdalsem) [inline mod/messageboard/views/default/resources/messageboard/owner.php:21]: sort by?

File: mod/messageboard/views/default/widgets/messageboard/content.php [hunk 29]
```diff
@@ -24,8 +24,8 @@
 	'limit' => $num_display,
 	'pagination' => false,
 	'order_by' => [
-		new OrderByClause('n_table.time_created', 'DESC'),
-		new OrderByClause('n_table.id', 'DESC'),
+		new OrderByClause('a_table.time_created', 'DESC'),
```
Comment (jdalsem) [inline mod/messageboard/views/default/widgets/messageboard/content.php:27]: sort by?

File: mod/pages/views/default/pages/sidebar/history.php [hunk 30]
```diff
@@ -17,12 +18,16 @@
 		'annotation_name' => 'page',
 		'limit' => max(20, elgg_get_config('default_limit')),
 		'order_by' => [
-			new \Elgg\Database\Clauses\OrderByClause('n_table.time_created', 'desc'),
-			new \Elgg\Database\Clauses\OrderByClause('n_table.id', 'desc'),
+			new OrderByClause('a_table.time_created', 'desc'),
```
Comment (jdalsem) [inline mod/pages/views/default/pages/sidebar/history.php:21]: sort by?

File: mod/pages/views/default/resources/pages/history.php [hunk 31]
```diff
@@ -21,8 +23,8 @@
 	'annotation_name' => 'page',
 	'limit' => max(20, elgg_get_config('default_limit')),
 	'order_by' => [
-		new \Elgg\Database\Clauses\OrderByClause('n_table.time_created', 'desc'),
-		new \Elgg\Database\Clauses\OrderByClause('n_table.id', 'desc'),
+		new OrderByClause('a_table.time_created', 'desc'),
```
Comment (jdalsem) [inline mod/pages/views/default/resources/pages/history.php:26]: sort by?

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/Database/Annotations.php [hunk 32]
```diff
@@ -141,7 +141,7 @@ public function get($limit = null, $offset = null, $callback = null) {
 	/**
 	 * Execute the query resolving calculation, count and/or batch options
 	 *
-	 * @return array|\ElggData[]|\ElggAnnotation[]|false|int
+	 * @return array|\ElggData[]|\ElggAnnotation[]|int|\ElggBatch
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/Annotations.php:144]: true

File: engine/classes/Elgg/Database/Annotations.php [hunk 33]
```diff
@@ -257,10 +257,10 @@ protected function buildPairedMetadataClause(QueryBuilder $qb, $clauses, $boolea
 		$parts = [];
 
 		foreach ($clauses as $clause) {
-			if (strtoupper($boolean) === 'OR' || count($clauses) > 1) {
-				$joined_alias = $qb->joinMetadataTable('n_table', 'entity_guid');
+			if (strtoupper($boolean) === 'OR' || count($clauses) === 1) {
+				$joined_alias = $qb->joinMetadataTable($qb->getTableAlias(), 'entity_guid', null, 'inner', MetadataTable::DEFAULT_JOIN_ALIAS);
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/Annotations.php:261]: will not work to get join aliases like `qbt1` and therefore will cause conflicts in queries

File: engine/classes/Elgg/Database/Annotations.php [hunk 34]
```diff
@@ -284,10 +284,10 @@ protected function buildPairedRelationshipClause(QueryBuilder $qb, $clauses, $bo
 
 		foreach ($clauses as $clause) {
 			$join_on = $clause->join_on == 'guid' ? 'entity_guid' : $clause->join_on;
-			if (strtoupper($boolean) == 'OR' || count($clauses) > 1) {
-				$joined_alias = $qb->joinRelationshipTable('n_table', $join_on, null, $clause->inverse);
+			if (strtoupper($boolean) === 'OR' || count($clauses) === 1) {
+				$joined_alias = $qb->joinRelationshipTable($qb->getTableAlias(), $join_on, null, $clause->inverse, 'inner', RelationshipsTable::DEFAULT_JOIN_ALIAS);
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/Annotations.php:288]: will not work to get join aliases like `qbt1` and therefore will cause conflicts in queries

File: engine/classes/Elgg/Database/Annotations.php [hunk 35]
```diff
@@ -214,7 +214,7 @@ protected function buildQuery(QueryBuilder $qb) {
 	 * @return \Closure|CompositeExpression|mixed|null|string
 	 */
 	protected function buildEntityWhereClause(QueryBuilder $qb) {
-		$joined_alias = $qb->joinEntitiesTable('n_table', 'entity_guid', 'inner', 'e');
+		$joined_alias = $qb->joinEntitiesTable($qb->getTableAlias(), 'entity_guid', 'inner', EntityTable::DEFAULT_JOIN_ALIAS);
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/Annotations.php:217]: will not work to get join aliases like `qbt1` and therefore will cause conflicts in queries

File: engine/classes/Elgg/Database/Entities.php [hunk 36]
```diff
@@ -322,10 +317,10 @@ protected function buildPairedRelationshipClause(QueryBuilder $qb, $clauses, $bo
 		$parts = [];
 
 		foreach ($clauses as $clause) {
-			if (strtoupper($boolean) == 'OR' || count($clauses) === 1) {
-				$joined_alias = $qb->joinRelationshipTable('e', $clause->join_on, null, $clause->inverse, 'inner', 'r');
+			if (strtoupper($boolean) === 'OR' || count($clauses) === 1) {
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/Entities.php:320]: nope, discussed

File: engine/classes/Elgg/Database/Metadata.php [hunk 37]
```diff
@@ -105,20 +105,20 @@ public function calculate($function, $property, $property_type = null) {
 	 * @return \ElggMetadata[]
 	 */
 	public function get($limit = null, $offset = null, $callback = null) {
-		$qb = Select::fromTable('metadata', 'n_table');
+		$qb = Select::fromTable(MetadataTable::TABLE_NAME, MetadataTable::DEFAULT_JOIN_ALIAS);
 
 		$distinct = $this->options->distinct ? 'DISTINCT' : '';
-		$qb->select("{$distinct} n_table.*");
+		$qb->select("{$distinct} {$qb->getTableAlias()}.*");
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/Metadata.php:111]: no idea,  fixed

File: engine/classes/Elgg/Database/RelationshipsTable.php [hunk 38]
```diff
@@ -25,7 +25,11 @@ class RelationshipsTable {
 	/**
 	 * @var integer The max length of the relationship column data
 	 */
-	const RELATIONSHIP_COLUMN_LENGTH = 255;
+	public const RELATIONSHIP_COLUMN_LENGTH = 255;
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/RelationshipsTable.php:28]: yes used in 
https://github.com/Elgg/Elgg/blob/7a6f5487461e3f99ac8263b5f8ebacf96af78b82/engine/tests/phpunit/unit/Elgg/Database/RelationshipsTableUnitTest.php#L57

File: engine/classes/Elgg/Database/Repository.php (hunk 1 above)
Comment (jeabakker) [reply engine/classes/Elgg/Database/Repository.php:187]: changed

File: mod/blog/views/default/blog/sidebar/revisions.php (hunk 25 above)
Comment (jeabakker) [reply mod/blog/views/default/blog/sidebar/revisions.php:22]: no

File: engine/classes/Elgg/Database/Metadata.php [hunk 39]
```diff
@@ -281,9 +281,9 @@ protected function buildPairedRelationshipClause(QueryBuilder $qb, $clauses, $bo
 
 		foreach ($clauses as $clause) {
 			if (strtoupper($boolean) == 'OR' || count($clauses) > 1) {
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/Metadata.php]: fixed

**Review by jdalsem — APPROVED** [review]

**Pull request merged**